packages/jsii/lib/assembler.ts

```diff
@@ -0,0 +1,879 @@
+            this._diagnostic(type.symbol.valueDeclaration, ts.DiagnosticCategory.Error, `Cannot use private type ${tsName} in exported declarations`);
+            return tsName;
+        }
+        const [, modulePath, typeName, ] = groups;
+        const pkg = await _findPackageInfo(modulePath);
+        if (!pkg) {
+            this._diagnostic(type.symbol.valueDeclaration, ts.DiagnosticCategory.Error, `Could not find module for ${modulePath}`);
+            return `unknown.${typeName}`;
+        }
+        const fqn = `${pkg.name}.${typeName}`;
+        if (pkg.name !== this.projectInfo.name && !this._dereference({ fqn })) {
+            this._diagnostic(type.symbol.valueDeclaration,
+                             ts.DiagnosticCategory.Error,
+                             `Use of foreign type not present in the ${pkg.name}'s assembly: ${fqn}`);
+        }
+        return fqn;
+
+        async function _findPackageInfo(fromDir: string): Promise<any> {
+            const filePath = path.join(fromDir, 'package.json');
+            if (await fs.pathExists(filePath)) {
+                return await fs.readJson(filePath);
+            }
+            const parent = path.dirname(fromDir);
+            if (parent === fromDir) { return undefined; }
+            return await _findPackageInfo(parent);
+        }
+    }
+
+    /**
+     * Register exported types in ``this.types``.
+     *
+     * @param node       a node found in a module
+     * @param namePrefix the prefix for the types' namespaces
+     */
+    private async _visitNode(node: ts.Declaration, namePrefix: string[] = []): Promise<spec.Type[]> {
+        // tslint:disable-next-line:no-bitwise Ignore nodes that are not export declarations
+        if ((ts.getCombinedModifierFlags(node) & ts.ModifierFlags.Export) === 0) { return []; }
+
+        let jsiiType: spec.Type | undefined;
+
+        if (ts.isClassDeclaration(node) && _isExported(node)) {
+            jsiiType = await this._visitClass(this._typeChecker.getTypeAtLocation(node), namePrefix);
+        } else if (ts.isInterfaceDeclaration(node) && _isExported(node)) {
+            jsiiType = await this._visitInterface(this._typeChecker.getTypeAtLocation(node), namePrefix);
+        } else if (ts.isEnumDeclaration(node) && _isExported(node)) {
+            jsiiType = await this._visitEnum(this._typeChecker.getTypeAtLocation(node), namePrefix);
+        } else if (ts.isModuleDeclaration(node)) {
+            const moduleType = this._typeChecker.getTypeAtLocation(node);
+            if (LOG.isTraceEnabled()) {
+                LOG.trace(`Entering namespace: ${colors.cyan([...namePrefix, moduleType.symbol.name].join('.'))}`);
+            }
+            const allTypes = new Array<spec.Type>();
+            for (const prop of moduleType.getProperties()) {
+                allTypes.push(...await this._visitNode(prop.valueDeclaration, namePrefix.concat(node.name.getText())));
+            }
+            if (LOG.isTraceEnabled()) {
```

Comment:
Why do you need to check if trace is enabled?

Reply:
Again, its a way to make compilation faster by not building large interpolated strings if we're not going to use them.